cabot/cabotapp/tests/test_schedules.py

```diff
@@ -138,3 +152,206 @@ def test_calendar_long_name(self):
         officers = get_duty_officers(self.schedule, at_time=datetime(2016, 11, 7, 10, 0, 0))
         emails = [str(user.email) for user in officers]
         self.assertEqual(emails, ['longnamelongnamelongnamelongname@affirm.com'])
+
+
+class TestScheduleValidation(LocalTestCase):
+    def setUp(self):
+        super(TestScheduleValidation, self).setUp()
+
+        _create_fake_users([
+            'dolores@affirm.com',
+            'bernard@affirm.com',
+            'teddy@affirm.com',
+            'maeve@affirm.com',
+            'hector@affirm.com',
+            'armistice@affirm.com',
+            'longnamelongnamelongnamelongname@affirm.com',
+            'shortname@affirm.com',
+        ])
+
+        self.secondary_schedule.delete()
+        self.secondary_schedule = None
+
+    @patch('cabot.cabotapp.models.requests.get', fake_calendar)
+    @patch('cabot.cabotapp.tasks.send_mail')
+    @patch('cabot.cabotapp.models.timezone.now')
+    def test_validate_schedule_no_gaps(self, fake_now, fake_send_mail):
+        fake_now.return_value = datetime(2018, 8, 15, 0, 3, 54, 598552, tzinfo=timezone.utc)
+
+        self.schedule.ical_url = 'calendar_no_gaps.ics'
+        self.schedule.fallback_officer = User.objects.get(email='dolores@affirm.com')
+        self.schedule.save()
+        tasks.update_shifts_and_problems()  # simulate periodic task firing
+        self.schedule = Schedule.objects.get(pk=self.schedule.pk)  # refresh from DB
+
+        self.assertFalse(fake_send_mail.called)
+        self.assertFalse(self.schedule.has_problems())
+
+    @patch('cabot.cabotapp.models.requests.get', fake_calendar)
+    @patch('cabot.cabotapp.tasks.send_mail')
+    @patch('cabot.cabotapp.models.timezone.now')
+    def test_validate_schedule_with_gaps(self, fake_now, fake_send_mail):
+        fake_now.return_value = datetime(2018, 8, 15, 0, 3, 54, 598552, tzinfo=timezone.utc)
+
+        self.schedule.ical_url = 'calendar_with_gaps.ics'
+        self.schedule.fallback_officer = User.objects.filter(email='dolores@affirm.com').first()
+        self.schedule.save()
+        tasks.update_shifts_and_problems()  # simulate periodic task firing
+        self.schedule = Schedule.objects.get(pk=self.schedule.pk)  # refresh from DB
+
+        message = """\
+The schedule <a href="{edit_schedule_url}">Principal</a> has some issues:
+
+There are gaps in the schedule:
+* 2018-08-17 19:00:00+00:00 to 2018-08-18 19:00:00+00:00
```

Comment:
Not necessary, but it might be nice to append an approximate duration at the end of each gap.  Eg:

```
2018-08-17 19:00:00+00:00 to 2018-08-18 19:00:00+00:00 (24 hours)
```

Reply:
Good call, that would be more usable and is easy to implement. Should clean up the timezone stuff here too (+00:00 with each timestamp is noisy, should probably get rid of seconds too).

Reply:
Added, times now look like this:

```
There are gaps in the schedule (times are UTC):
* 2018-08-15 00:03:54.598552 to 2018-08-21 17:53:28 (6d, 17h, 49m, 33s)
* 2018-08-21 23:45:16 to 2018-08-22 07:00:00 (7h, 14m, 44s)
* 2018-08-22 08:00:00 to 2018-08-23 07:00:00 (23h)
* 2018-08-23 08:00:00 to 2018-08-24 07:00:00 (23h)
* 2018-08-24 08:00:00 to 2018-08-25 07:00:00 (23h)
```